### PR TITLE
Add support for GHC 9.4.1

### PR DESCRIPTION
**Pull request by christiaanb**
This is work in progress

### CODE REVIEW

**Comment by gelisam** [comment]
I really liked how @brandon-leapyear did it [last time](https://github.com/haskell-hint/hint/pull/139/files#diff-1fb5582a6e6e6976c911506aaeb976332fb7ff173f4416a6cca237345faac991R217): instead of adding `ifdef`s all over the codebase, he put them all inside _shims_. Each shim is a wrapper around a particular GHC function, with a type signature which is carefully picked so that it can be implemented with every supported GHC version, sometimes  as a no-op or by ignoring some of its arguments. This way, the rest of the library implementation can be written against the shim's API, without having to worry about the GHC version.

That being said, it is much more important to get the library to work against GHC-9.4.1 than to make the code look nice, so perhaps you can focus on making it work and then I can worry about refactoring the code to the shims style :)

**Comment by christiaanb** [comment]
Yeah, the plan is to make it work first.

Currently everything compiles, but none of the test pass because it doesn’t load the modules. Once that’s sorted I’ll refactor to use shims.

**Comment by brandonchinn178** [comment]
FYI I found this other technique to minimize the number of #if macros even more, if you're interested in it: https://github.com/brandonchinn178/tasty-autocollect/blob/main/src/Test/Tasty/AutoCollect/GHC/Shim.hs

https://www.reddit.com/r/haskell/comments/2uehlo/stop_abusing_cpp_in_haskell_sources/?utm_medium=android_app&utm_source=share

**Comment by gelisam** [comment]
I had to add the following `cabal.project` file in order to get `cabal test` to run. I assume you have something similar locally as well. Should we commit this to the repo, in order to help others compile with ghc-9.4.1? Maybe under the name `cabal.project-ghc-9.4.1`, so that these version bound overrides don't get used in the common case in which users are using a released version of ghc?

```
packages: .
tests: True
constraints:
  async >= 2.0.0.0
allow-newer:
  async:base,
  hashable:base,
  hashable:ghc-bignum
```

**Comment by christiaanb** [comment]
I don’t know, once those packages are upgraded the `cabal.project` file would be redundant.

**Comment by martijnbastiaan** [comment]
@christiaanb I doubt you have time to address this, but do you have any sense of what's wrong / a way to fix it you could share with us?

**Comment by christiaanb** [comment]
Right, the issue is that this context storing and restoring: https://github.com/haskell-hint/hint/blob/42afe8197e7cc8b68d0ba005b42efdc027c95b57/src/Hint/Context.hs#L126-L134

is no longer sufficient. And I haven't figure out how to implement it in the GHC 9.4 API.

This means that https://github.com/haskell-hint/hint/blob/42afe8197e7cc8b68d0ba005b42efdc027c95b57/src/Hint/Context.hs#L414-L419 causes the non-Phantom modules to be unloaded. Which in turn means that basically we only have the Phantom modules in scope, which are rather useless by themselves.

**Comment by danwdart** [comment]
At the moment with this, I got unit test failures on ghc 9.4.2:

```
running tests
Running 1 test suites...
Test suite unit-tests: RUNNING...
^MCases: 16  Tried: 0  Errors: 0  Failures: 0^M                                           ^M### Failure in: 0:reload_modified
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nTEST_ReloadModified\n"
^MCases: 16  Tried: 1  Errors: 0  Failures: 1^MCases: 16  Tried: 2  Errors: 0  Failures: 1^M                                           ^M### Failure in: 2:work_in_main
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nMain\n"
^MCases: 16  Tried: 3  Errors: 0  Failures: 2^MCases: 16  Tried: 4  Errors: 0  Failures: 2^M                                           ^M### Failure in: 4:qual_import
unit-tests/run-unit-tests.hs:471
GhcException "Cannot add module M68944334256838729282258 to context: not a home module"
^MCases: 16  Tried: 5  Errors: 0  Failures: 3^M                                           ^M### Failure in: 5:full_import
unit-tests/run-unit-tests.hs:471
GhcException "Cannot add module M77412322302094290192258 to context: not a home module"
^MCases: 16  Tried: 6  Errors: 0  Failures: 4^MCases: 16  Tried: 7  Errors: 0  Failures: 4^MCases: 16  Tried: 8  Errors: 0  Failures: 4^MCases: 16  Tried: 9  Errors: 0  Failures: 4^MCases>
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nT\n"
^MCases: 16  Tried: 11  Errors: 0  Failures: 5^MCases: 16  Tried: 12  Errors: 0  Failures: 5^MCases: 16  Tried: 13  Errors: 0  Failures: 5^MCases: 16  Tried: 14  Errors: 0  Failures: 5^M >
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nMain\n"
^MCases: 16  Tried: 15  Errors: 0  Failures: 6^M                                            ^M### Failure in: 15:normalize_type
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nT\n"
Cases: 16  Tried: 16  Errors: 0  Failures: 7
^MCases: 2  Tried: 0  Errors: 0  Failures: 0^MCases: 2  Tried: 1  Errors: 0  Failures: 0^M                                          ^M### Error in:   1:package_db
cabal: startProcess: exec: does not exist (No such file or directory)
Cases: 2  Tried: 2  Errors: 1  Failures: 0
^MCases: 16  Tried: 0  Errors: 0  Failures: 0^M                                           ^M### Failure in: 0:reload_modified
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nTEST_ReloadModified\n"
^MCases: 16  Tried: 1  Errors: 0  Failures: 1^MCases: 16  Tried: 2  Errors: 0  Failures: 1^M                                           ^M### Failure in: 2:work_in_main
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nMain\n"
^MCases: 16  Tried: 3  Errors: 0  Failures: 2^MCases: 16  Tried: 4  Errors: 0  Failures: 2^M                                           ^M### Failure in: 4:qual_import
unit-tests/run-unit-tests.hs:471
GhcException "Cannot add module M66940689969892995022258 to context: not a home module"
^MCases: 16  Tried: 5  Errors: 0  Failures: 3^M                                           ^M### Failure in: 5:full_import
unit-tests/run-unit-tests.hs:471
GhcException "Cannot add module M58670291326063233382258 to context: not a home module"
^MCases: 16  Tried: 6  Errors: 0  Failures: 4^MCases: 16  Tried: 7  Errors: 0  Failures: 4^MCases: 16  Tried: 8  Errors: 0  Failures: 4^MCases: 16  Tried: 9  Errors: 0  Failures: 4^MCases>
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nT\n"
^MCases: 16  Tried: 11  Errors: 0  Failures: 5^MCases: 16  Tried: 12  Errors: 0  Failures: 5^MCases: 16  Tried: 13  Errors: 0  Failures: 5^MCases: 16  Tried: 14  Errors: 0  Failures: 5^M >
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nMain\n"
^MCases: 16  Tried: 15  Errors: 0  Failures: 6^M                                            ^M### Failure in: 15:normalize_type
unit-tests/run-unit-tests.hs:471
NotAllowed "These modules have not been loaded:\nT\n"
Cases: 16  Tried: 16  Errors: 0  Failures: 7
^MCases: 2  Tried: 0  Errors: 0  Failures: 0^MCases: 2  Tried: 1  Errors: 0  Failures: 0^M                                          ^M### Error in:   1:package_db
cabal: startProcess: exec: does not exist (No such file or directory)
Cases: 2  Tried: 2  Errors: 1  Failures: 0
Test suite unit-tests: FAIL
```

**Comment by danwdart** [comment]
Otherwise the package seems to compile without running these.

**Comment by yaxu** [comment]
Hi, is ghc 9.4 support on the horizon for hint?

**Comment by gelisam** [comment]
I have not had much time for open source projects recently, but thanks for reminding me about this, it helps me to prioritize which issues to focus on when I do find the time!

**Comment by martijnbastiaan** [comment]
This seems to come down to behavioral changes in `load`, I've reported an issue over here: https://gitlab.haskell.org/ghc/ghc/-/issues/23353. AFAICT one of these options need to happen:

 1. Hint has to change to not rely on this behavior
 2. GHC needs to revert to old behavior
 3. GHC needs to add an additional load option `LoadAllTargetsUpTo` (see issue report)

**Comment by gelisam** [comment]
Thanks for tracking this down! However, there is still something I don't understand. Immediately above the code where hint calls LoadUpTo, there is [a comment](https://github.com/haskell-hint/hint/blob/7ba9700a8a1f6d3bd5aeb32fde4e1eab61601b04/src/Hint/Context.hs#L126) which says:

> GHC.load will remove all the modules from scope, so first we save the context

Then there are calls to `getContext` and `setContext` around `GHC.load`. This seems to imply that:

1. hint already expects `GHC.load` to unload all the modules, and even already has a workaround.
2. The workaround doesn't work anymore.

That doesn't make much sense though, since your test case demonstrates that ghc's behavior has changed. Maybe an even older version of GHC was behaving like GHC 9.4 in that regard? Or maybe the comment isn't saying that the modules get _unloaded_, instead it is saying that they are still loaded but that they need to be imported again?

**Comment by martijnbastiaan** [comment]
> Or maybe the comment isn't saying that the modules get _unloaded_, instead it is saying that they are still loaded but that they need to be imported again?

I'm pretty sure this is it. There's still a chance I'm chasing a red herring though, so I'll implement a patch for GHC and see if it fixes Hint's issues.

**Comment by gelisam** [comment]
> I'll implement a patch for GHC and see if it fixes Hint's issues.

That sounds like a lot of work! Personally, I would start with approach 1, trying to change Hint to not rely on this behavior. After all, we know exactly what modules should be in scope, so it shouldn't be too hard to list them all if that's what the new GHC API requires.

**Comment by martijnbastiaan** [comment]
> That sounds like a lot of work!

It's quite alright, I found the process relatively straightforward. I implemented two patches; one fixing a straight up bug in `load`, another for the desired new constructor: https://gitlab.haskell.org/martijnbastiaan/ghc/-/commits/add-load-all-targets-upto-9.4.5/.

Unfortunately, the trick:

```
(old_top, old_imps) <- getContext
-- do stuff
setContext old_top old_imps
```

doesn't work anymore for newer GHCs, presumably because references aren't valid after calling `load` again. So it looks like Hint will have to change its behavior.

**Comment by yaxu** [comment]
Excellent news thanks so much @christiaanb !

File: src/Hint/Context.hs [hunk 1]
```diff
@@ -128,7 +137,11 @@ addPhantomModule mod_text =
                    (old_top, old_imps) <- runGhc getContext
                    --
                    runGhc $ GHC.addTarget t
+#if MIN_VERSION_ghc(9,4,0)
+                   res <- runGhc $ GHC.load GHC.LoadAllTargets
+#else
                    res <- runGhc $ GHC.load (GHC.LoadUpTo m)
+#endif
```
Comment (martijnbastiaan) [inline src/Hint/Context.hs:144]: Couldn't we use `LoadAllTargets` for all GHC versions now?

File: unit-tests/run-unit-tests.hs [hunk 2]
```diff
@@ -318,7 +318,7 @@ test_package_db = IOTestCase "package_db" [dir] $ \wrapInterp -> do
               & last
                 -- "8.8.4"
         let pkgdb    = dir </> "dist-newstyle" </> "packagedb" </> ("ghc-" ++ ghcVersion)
-            ghc_args = ["-package-db=" ++ pkgdb]
+            ghc_args = ["-package-db=" ++ pkgdb,"-package-env","-"]
```
Comment (martijnbastiaan) [inline unit-tests/run-unit-tests.hs:321]: I'm curious: did you enable `write-ghc-environment-variables` globally in your by any chance? It seems like this flag would only be necessary if you set `GHC_ENVIRONMENT` or have an environment file lying around.

Comment (christiaanb) [reply unit-tests/run-unit-tests.hs:321]: Ah, yeah, you're right. This change won't be needed.

File: src/Hint/Context.hs (hunk 1 above)
Comment (christiaanb) [reply src/Hint/Context.hs:144]: Perhaps... I guess I can try on the oldest supported version.

File: unit-tests/run-unit-tests.hs (hunk 2 above)
Comment (gelisam) [reply unit-tests/run-unit-tests.hs:321]: Thus reminds me that I really need to improve the documentation of how hint finds its package databases. The easy way to use hint is to build with `write-ghc-environment-files` enabled, this way hint can automatically find its package databases at runtime. The hard way is to figure out where cabal and stack hide their package databases, and to specify a path to them via these `-package-db` args.

Since this test is specifically about checking that hint can use this hard way, it makes sense to disable the environment file in this test using `-package-env -`. At some point it would be nice to also add a test for the easy way!

**Comment by christiaanb** [comment]
@gelisam I've cleaned up the PR, could you give it another review, and if everything looks good to you merge the PR into `main`?

File: src/Hint/Context.hs [hunk 3]
```diff
@@ -128,7 +125,7 @@ addPhantomModule mod_text =
                    (old_top, old_imps) <- runGhc getContext
                    --
                    runGhc $ GHC.addTarget t
-                   res <- runGhc $ GHC.load (GHC.LoadUpTo m)
+                   res <- runGhc $ GHC.loadPhantomModule m
```
Comment (samuel-gelineau-at-well-dot) [inline src/Hint/Context.hs:128]: I don't understand. With ghc-9.4, `loadPhantomModule m` is defined as `GHC.load GHC.LoadAllTargets`, which does not mention `m`. So where does `m` get loaded?

Comment (samuel-gelineau-at-well-dot) [reply src/Hint/Context.hs:128]: Also, [this comment](https://github.com/haskell-hint/hint/pull/152#issuecomment-1537520685) says that calling `getContext` and then `setContext` does not work anymore, because the saved context refers to the modules using some internal IDs which are no longer valid after calling `LoadAllTargets`. So why is the code still using `getContext` and `setContext`? Is the comment wrong, and `setContext` does work after all?

File: src/Hint/Context.hs [hunk 4]
```diff
@@ -385,14 +380,27 @@ reset = do -- clean up context
...
+                                Nothing -> do
+                                  mod' <- findModule (pmName mod)
+                                  runGhc $ setContext [mod'] []
+                                  return GHC.Succeeded
+                                Just ts -> do
+                                  runGhc $ GHC.setTargets ts
+                                  df <- runGhc GHC.getSessionDynFlags
+                                  let t = GHC.fileTarget df (pmFile mod)
+                                  runGhc $ GHC.addTarget t
+                                  res <- runGhc $ GHC.load GHC.LoadAllTargets
```
Comment (samuel-gelineau-at-well-dot) [inline src/Hint/Context.hs:400]: I am confused. `addPhantomModule` is basically running those two lines, `addTarget` and then `LoadAllTargets`. So why do we need to reset the targets and then run those two lines again after `addPhantomModule` returns? Could we maybe call `setTargets` before calling `addPhantomModule` instead?

File: src/Hint/Context.hs [hunk 5]
```diff
@@ -385,14 +380,27 @@ reset = do -- clean up context
...
-                          mod' <- findModule (pmName mod)
-                          runGhc $ setContext [mod'] []
+installSupportModule :: MonadInterpreter m => Maybe [GHC.Target] -> m GHC.SuccessFlag
+installSupportModule tM = do mod <- addPhantomModule support_module
+                             onState (\st -> st{hintSupportModule = mod})
+                             case tM of
+                                Nothing -> do
+                                  mod' <- findModule (pmName mod)
+                                  runGhc $ setContext [mod'] []
+                                  return GHC.Succeeded
```
Comment (samuel-gelineau-at-well-dot) [inline src/Hint/Context.hs:394]: I now understand that this `Nothing` case runs if we try and fail to load the modules which the user has requested. If the exception propagates all the way to the outside of the InterpreterT block, then it doesn't really matter what is in the context, but if the user catches the exception inside the InterpreterT block and then continues to evaluate code, it makes sense that we now want the context to be empty except for the support module.

Because this is such a rare corner case, I am worried about whether we have a test for that case. This comment is a reminder for me to check if we have a test case, and to add a new one otherwise.

File: src/Hint/Context.hs [hunk 6]
```diff
@@ -128,7 +125,7 @@ addPhantomModule mod_text =
                    (old_top, old_imps) <- runGhc getContext
```
Comment (samuel-gelineau-at-well-dot) [inline src/Hint/Context.hs:125]: The comment above mentions `GHC.load`, which is no longer present in this PR's code.

**Review by gelisam — APPROVED** [review]

**Comment by gelisam** [comment]
Sorry it took me so long to find the time to review this!

Since you obviously care about the library and you clearly have more time to dedicate to it than I do, would you like to take over as maintainer?

File: src/Hint/Context.hs (hunk 3 above)
Comment (gelisam) [reply src/Hint/Context.hs:128]: I now think that `setContext` does fail, but then the caller fixes it.

File: src/Hint/InterpreterT.hs [hunk 7]
```diff
@@ -181,26 +181,9 @@ newSessionData a =
...
-#if MIN_VERSION_ghc(9,0,0)
-            \df wr sev src msg -> f df wr sev src id msg
-#else
-            \df wr sev src style msg -> f df wr sev src (GHC.withPprStyle style) msg
-#endif
-
-mkGhcError :: (GHC.SDoc -> String) -> GHC.SrcSpan -> GHC.Message -> GhcError
-mkGhcError render src_span msg = GhcError{errMsg = niceErrMsg}
-    where niceErrMsg = render $ GHC.mkLocMessage GHC.SevError src_span msg
-
```
Comment (gelisam) [inline src/Hint/InterpreterT.hs:203]: Thanks for moving all this stuff to the stubs file!

**Comment by gelisam** [comment]
> Unfortunately, the trick:
> 
>     (old_top, old_imps) <- getContext
>     -- do stuff
>     setContext old_top old_imps
> 
> doesn't work anymore for newer GHCs, presumably because references aren't valid after calling load again. So it looks like Hint will have to change its behavior.

@martijnbastiaan what made you conclude that this trick doesn't work anymore? It seems on the contrary that the trick works and [is essential](https://github.com/haskell-hint/hint/pull/152/files#r1231273589) for the `test_work_in_main` test to pass.

File: src/Hint/Context.hs (hunk 5 above)
Comment (gelisam) [reply src/Hint/Context.hs:394]: test added in af97170ad223ccca48b539484bfc442cf9641884

File: unit-tests/run-unit-tests.hs (hunk 2 above)
Comment (gelisam) [reply unit-tests/run-unit-tests.hs:321]: documentation added in ff15999fee2241b41858907ae5645eb1e95dc176

test added in dad34b9685d14aea4335ba8eed63532eaaa4f384

**Comment by gelisam** [comment]
I now think that the main change in ghc-9.4 isn't to `getContext` and `setContext`, but rather to `LoadUpTo m`. Previously, it kept all currently-loaded targets and also loaded `m` and its dependencies. Now, it unloads all currently-loaded targets and _only_ loads `m` and its dependencies. Thus, we now need to call `LoadAllTargets` instead of `LoadUpTo`, which is what this PR does.

File: src/Hint/Context.hs (hunk 4 above)
Comment (gelisam) [reply src/Hint/Context.hs:400]: 567c057bbb89f5b694ce0a0881ee2b77b104317c now calls `setTargets` before calling `addPhantomModule`

**Comment by gelisam** [comment]
> All that remains is some cleaning up and we should be able to make a new release.

released as [hint-0.9.0.7](https://hackage.haskell.org/package/hint-0.9.0.7)

**Comment by martijnbastiaan** [comment]
> @martijnbastiaan what made you conclude that this trick doesn't work anymore? It seems on the contrary that the trick works and [is essential](https://github.com/haskell-hint/hint/pull/152/files#r1231273589) for the test_work_in_main test to pass.

I believe it - I probably got confused at some point while hacking on Hint.
